src/lang/std/artifactGraph.ts

```diff
@@ -0,0 +1,645 @@
+  pathId: string
+  surfs: Array<_WallArtifact | _CapArtifact>
+  edges: Array<string>
+  codeRef: CommonCommandProperties
+}
+
+interface _WallArtifact {
+  type: 'wall'
+  segId: string
+  edgeCutEdgeIds: Array<string>
+  extrusionId: string
+  pathIds: Array<string>
+}
+interface _CapArtifact {
+  type: 'cap'
+  subType: 'start' | 'end'
+  edgeCutEdgeIds: Array<string>
+  extrusionId: string
+  pathIds: Array<string>
+}
+
+interface _ExtrudeEdge {
+  type: 'extrudeEdge'
+  segId: string
+  extrusionId: string
+  edgeId: string
+}
+
+/** A edgeCut is a more generic term for both fillet or chamfer */
+interface _EdgeCut {
+  type: 'edgeCut'
+  subType: 'fillet' | 'chamfer'
+  consumedEdgeId: string
+  edgeIds: Array<string>
+  surfId: string
+  codeRef: CommonCommandProperties
+}
+
+interface EdgeCutEdge {
+  type: 'edgeCutEdge'
+  edgeCutId: string
+  surfId: string
+}
+
+export type Artifact =
+  | _PlaneArtifact
+  | _PathArtifact
+  | _SegmentArtifact
+  | _ExtrusionArtifact
+  | _WallArtifact
+  | _CapArtifact
+  | _ExtrudeEdge
+  | _EdgeCut
+  | EdgeCutEdge
+  | _solid2D
+
+export type ArtifactGraph = Map<string, Artifact>
+
+export type EngineCommand = Models['WebSocketRequest_type']
+
+type OkWebSocketResponseData = Models['OkWebSocketResponseData_type']
+
+export interface ResponseMap {
+  [commandId: string]: OkWebSocketResponseData
+}
+export interface OrderedCommand {
+  command: EngineCommand
+  range: SourceRange
+}
+
+export function createArtifactGraph({
+  orderedCommands,
+  responseMap,
+  ast,
+}: {
+  orderedCommands: Array<OrderedCommand>
+  responseMap: ResponseMap
+  ast: Program
+}) {
+  const myMap = new Map<string, Artifact>()
+  let currentPlaneId = ''
+
+  const getArtifact = (id: string) => myMap.get(id)
+  orderedCommands.forEach((orderedCommand) => {
+    if (orderedCommand.command?.type === 'modeling_cmd_req') {
+      if (orderedCommand.command.cmd.type === 'enable_sketch_mode') {
+        currentPlaneId = orderedCommand.command.cmd.entity_id
+      }
+      if (orderedCommand.command.cmd.type === 'sketch_mode_disable') {
+        currentPlaneId = ''
+      }
+    }
+    const modArr = getArtifactsToUpdate({
+      orderedCommand,
+      responseMap,
+      getArtifact,
+      currentPlaneId,
+      ast,
+    })
+    modArr.forEach(({ id, artifact }) => {
+      const oldArtifact = myMap.get(id)
+      if (oldArtifact?.type !== artifact.type) {
+        myMap.set(id, artifact)
+        return
+      }
+      const mergedArtifact = { ...oldArtifact, ...artifact }
+      Object.entries(artifact).forEach(([propName, value]) => {
+        const otherValue = (oldArtifact as any)[propName]
+        if (Array.isArray(value) && Array.isArray(otherValue)) {
+          ;(mergedArtifact as any)[propName] = [
+            ...new Set([...otherValue, ...value]),
+          ]
+        }
+      })
+      myMap.set(id, mergedArtifact)
+    })
+  })
+  return myMap
+}
+
+export function getArtifactsToUpdate({
```

Comment:
I didn't want this function to mutate the map at all, only to return info on how to mutate the map.